ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/validate/DeduplicationState.scala

```diff
@@ -51,15 +53,80 @@ case class DeduplicationState private (
 }
 
 object DeduplicationState {
-  private[sandbox] type DeduplicationQueue = VectorMap[ChangeId, Time.Timestamp]
+  private[sandbox] type DeduplicationQueue = UnsafeDeduplicationStateQueueMap
 
   private[validate] def empty(
       deduplicationDuration: Duration,
       bridgeMetrics: BridgeMetrics,
   ): DeduplicationState =
     DeduplicationState(
-      deduplicationQueue = VectorMap.empty,
+      deduplicationQueue = UnsafeDeduplicationStateQueueMap.empty,
       maxDeduplicationDuration = deduplicationDuration,
       bridgeMetrics = bridgeMetrics,
     )
+
+  /** This data structure is tailored for keeping an ordered (by insertion) sequence of deduplication entries
+    * of the form (changeId, recordTime) with optimal complexity of each exposed operation.
+    *
+    * @param vector An ordered (by insertion) vector of deduplication entries
+    * @param mappings Mapping of changeId to recordTime
+    */
+  private[validate] case class UnsafeDeduplicationStateQueueMap(
+      vector: Vector[(ChangeId, Time.Timestamp)],
```

Comment:
AFAICS this is executed in a sequential state, possible improvement could be to use mutable queue and map in here, which would result in faster code. (optimization)

Reply:
The entire logic in the `SequenceImpl` revolves around the immutability guarantee of these states (also see `SequenceState`) and making everything mutable will complicate reasoning around that piece of code. So, I've preferred this approach since it is cleaner and readable while also idiomatic. On the other hand, as currently benchmarked, it seems we still have 6-8x headroom in computations, so I'm wary of over-optimizing at this point, while I'm also not sure how much we would gain (note that our data structures are still append- and drop-based so I wouldn't rely on in-memory value swaps too much). If you think it could make a sizeable difference, let me know; happy to spend more cycles on it.